### PR TITLE
[#243] enable RESC_HIER_STR_KW and RESC_NAME_KW

### PR DESCRIPTION
**Pull request by d-w-moore**

### CODE REVIEW

File: irods/manager/data_object_manager.py [hunk 1]
```diff
@@ -290,9 +290,15 @@ def open_with_FileRaw(self, *arg, **kw_options):
...
+    _RESC_flags_for_open = set(( kw.RESC_NAME_KW,
+                                 kw.DEST_RESC_NAME_KW,  # may be deprecated in the future -- TODO new issue
+                                 kw.RESC_HIER_STR_KW,
+                                 # kw.DEST_RESC_HIER_STR_KW,
+                               ))
+
     def open(self, path, mode, create = True, finalize_on_close = True, **options):
         _raw_fd_holder =  options.get('_raw_fd_holder',[])
-        if kw.DEST_RESC_NAME_KW not in options:
+        if 0 == len(self._RESC_flags_for_open & set(options.keys())):
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:301]: Python uses "&" (intersection) and "|" (union) as set operators.

[the remark this answers is not in the document]
Comment (alanking) [reply irods/manager/data_object_manager.py:301]: Didn't know about that, but I think it's pretty neat. For reference: https://docs.python.org/3/library/stdtypes.html?highlight=set#frozenset.intersection

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py]: Should be a bit better now.

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py]: Set this as non-draft. We'll take a look tomorrow.

**Comment by d-w-moore** [comment]
Just set this to non-draft this morning. Let's take a final look tomorrow....

File: irods/manager/data_object_manager.py [hunk 2]
```diff
@@ -290,9 +290,18 @@ def open_with_FileRaw(self, *arg, **kw_options):
         handle = self.open(*arg,_raw_fd_holder=holder,**kw_options)
         return (handle, holder[-1])
 
+    _RESC_flags_for_open = frozenset((
+                               kw.RESC_NAME_KW,
+                               kw.DEST_RESC_NAME_KW,  # may be deprecated in the future -- TODO new issue
+                               kw.RESC_HIER_STR_KW,
+                               # kw.DEST_RESC_HIER_STR_KW,
```
Comment (trel) [inline irods/manager/data_object_manager.py:297]: leave a note to say this is commented out on purpose
OR
remove this line

File: irods/test/data_obj_test.py [hunk 3]
```diff
@@ -1517,6 +1517,33 @@ def enter_file_into_irods( session, filename, **kw_opt ):
...
+                 self.create_simple_resc('resc1_243') as r1:
+                s.resources.add_child(parent.name,r0)
+                s.resources.add_child(parent.name,r1)
+                hiers = ['parent;{0}'.format(r) for r in (r0,r1)]
+                # write two different replicas
+                for hier in hiers:
+                    opts = {kw.RESC_HIER_STR_KW: hier}
+                    with s.data_objects.open(data_path, 'a', **opts) as obj_io:
+                        obj_io.seek(0)
+                        obj_io.write(hier.encode('utf-8'))   # Write different content to each replica
```
Comment (trel) [inline irods/test/data_obj_test.py:1536]: note in a comment also that the first replica will end up 'stale' and the second replica will be 'good'.

File: irods/manager/data_object_manager.py [hunk 4]
```diff
@@ -290,9 +290,18 @@ def open_with_FileRaw(self, *arg, **kw_options):
         handle = self.open(*arg,_raw_fd_holder=holder,**kw_options)
         return (handle, holder[-1])
 
+    _RESC_flags_for_open = frozenset((
+                               kw.RESC_NAME_KW,
+                               kw.DEST_RESC_NAME_KW,  # may be deprecated in the future -- TODO new issue
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:295]: Yes, will do.

File: irods/manager/data_object_manager.py [hunk 5]
```diff
@@ -290,9 +290,17 @@ def open_with_FileRaw(self, *arg, **kw_options):
         handle = self.open(*arg,_raw_fd_holder=holder,**kw_options)
         return (handle, holder[-1])
 
+    _RESC_flags_for_open = frozenset((
+                               kw.RESC_NAME_KW,
+                               kw.DEST_RESC_NAME_KW,  # may be deprecated in the future
+                               kw.RESC_HIER_STR_KW,
```
Comment (alanking) [inline irods/manager/data_object_manager.py:296]: Extra comma here. Not sure if that's on purpose

File: irods/test/data_obj_test.py [hunk 6]
```diff
@@ -1517,6 +1517,58 @@ def enter_file_into_irods( session, filename, **kw_opt ):
...
+                    f.write(b'abc')
+                d = data.get(data_path)
+                self.assertEqual(len(d.replicas),1)
+                self.assertEqual(d.replicas[0].resource_name, resc)
+            finally:
+                if data.exists(data_path):
+                    data.unlink(data_path, force = True)
+
+
+    def test_access_through_resc_hierarchy_243(self):
```
Comment (alanking) [inline irods/test/data_obj_test.py:1540]: Should this be `test_access_through_resc_hierarchy__243`?

File: irods/test/data_obj_test.py [hunk 7]
```diff
@@ -1517,6 +1517,58 @@ def enter_file_into_irods( session, filename, **kw_opt ):
...
+        data_name = unique_name(my_function_name(),datetime.now(),'data')
+        resc_name = unique_name(my_function_name(),datetime.now(),'resc')
+        with self.create_simple_resc(resc_name) as resc:
+            data_path = '{home}/{data_name}'.format(**locals())
+            try:
+                with data.open(data_path,'w',**{kw.RESC_NAME_KW:resc}) as f:
+                    f.write(b'abc')
+                d = data.get(data_path)
+                self.assertEqual(len(d.replicas),1)
+                self.assertEqual(d.replicas[0].resource_name, resc)
```
Comment (alanking) [inline irods/test/data_obj_test.py:1534]: Not sure how much of this is testing the client versus testing the server, but maybe you could target a non-root resource with this keyword and ensure that the right error message/code is returned? Should give you a `DIRECT_CHILD_ACCESS`, I think.

**Review by alanking — COMMENTED** [review]
Do we have a test which uses `DEST_RESC_HIER_STR_KW` (or any other unsupported keyword, for that matter) to demonstrate what happens when it is used with `open`? Does that add any value?

File: irods/manager/data_object_manager.py (hunk 5 above)
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:296]: > Do we have a test which uses `DEST_RESC_HIER_STR_KW` (or any other unsupported keyword, for that matter) to demonstrate what happens when it is used with `open`? Does that add any value?

The parallel put test uses DEST_RESC_HIER_STR_KW.

Comment (d-w-moore) [reply irods/manager/data_object_manager.py:296]: > Do we have a test which uses `DEST_RESC_HIER_STR_KW` (or any other unsupported keyword, for that matter) to demonstrate what happens when it is used with `open`? Does that add any value?

The present behavior is in the finest of PRC traditions, that is:  Let users apply the keywords they want to - and if the server ignores them silently,  the client will follows suit and the file ends up on the default resource.

Is that the behavior we want?  And do we care about its enforcement in the future?

Perhaps we'd prefer to chase down unfortunate keyword choices and throw exceptions when we see them.  If so, I expect we'd want to test for the success of trapping such a mistake, yes?

I guess the value of such a test depends on the resolution of these further questions.

File: irods/test/data_obj_test.py (hunk 7 above)
Comment (d-w-moore) [reply irods/test/data_obj_test.py:1534]: I guess that would be valuable, yes.  Not to mention DIRECT_CHILD_ACCESS seems ... not to be defined  in `irods.exceptions`. That would need fixing too.

File: irods/manager/data_object_manager.py (hunk 5 above)
Comment (alanking) [reply irods/manager/data_object_manager.py]: In light of those questions, I think adding such a test would not provide much value. Let's just leave it for now. Thanks!

**Comment by d-w-moore** [comment]
all conversations resolved, and codacy happy as i can make it. maximally squashed.

**Review by alanking — APPROVED** [review]
Please await one more approval and we can add #s

File: irods/test/data_obj_test.py [hunk 8]
```diff
@@ -1517,6 +1518,66 @@ def enter_file_into_irods( session, filename, **kw_opt ):
...
+
+                # Assert that we are able to read both replicas' content faithfully using the hierarchy string.
+                for hier in hiers:
+                    opts = {kw.RESC_HIER_STR_KW: hier}
+                    with s.data_objects.open(data_path, 'r', **opts) as obj_io:
+                        self.assertEqual(obj_io.read(),hier.encode('utf-8'))   # Does unique replica have unique content?
+
+                s.data_objects.unlink(data_path, force = True)
+                s.resources.remove_child(parent.name,r1)
+                s.resources.remove_child(parent.name,r0)
```
Comment (korydraughn) [inline irods/test/data_obj_test.py:1577]: Could be in a `finally` block?

**Review by korydraughn — COMMENTED** [review]
Looks good.

Found one thing.

File: irods/test/data_obj_test.py (hunk 8 above)
Comment (d-w-moore) [reply irods/test/data_obj_test.py]: Yes, or - even better (or at least, less confusingly) - use an RAII-like helper class.

**Comment by trel** [comment]
happy with the code as is.     should the helper class be its own issue/commit?      definitely fine if not.

**Review by trel — APPROVED** [review]

**Review by korydraughn — APPROVED** [review]
I'm fine with leaving the code as is. I don't think separating the helper into a different commit adds much value since it appears to be specific to the test file where it is used.

**Comment by d-w-moore** [comment]
I'm easy but still not clear on the reference point for "as is". That is, whether we want to scrap the "adopter" helper.

Lmk, then I can either dispose of that latest commit, or squash it in.

**Comment by korydraughn** [comment]
We're talking about whether the latest commit should get a dedicated issue or be squashed.

I'm fine with either, but squashing does seem more correct as the second commit completes the test IMO.

**Comment by trel** [comment]
agreed - squash it.

**Comment by d-w-moore** [comment]
squashed; codacy appeased as much as possible

**Review by trel — APPROVED** [review]

**Review by alanking — APPROVED** [review]

**Comment by alanking** [comment]
Sorry for pre-mature merge... Linking issues as appropriate...

Addresses https://github.com/irods/python-irodsclient/issues/243

**Pull request merged**